server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -651,6 +654,105 @@ public void testResetDesiredBalanceOnNoLongerMaster() {
+        var desiredBalanceAllocator = new DesiredBalanceShardsAllocator(
+            delegateAllocator,
+            threadPool,
+            clusterService,
+            desiredBalanceComputer,
+            (reconcilerClusterState, rerouteStrategy) -> reconcilerClusterState,
+            TelemetryProvider.NOOP
+        ) {
+            @Override
+            public void resetDesiredBalance() {
+                resetCalled.set(true);
+                super.resetDesiredBalance();
+            }
+        };
+
+        var service = createAllocationService(desiredBalanceAllocator, createGatewayAllocator());
+
+        try {
+            rerouteAndWait(service, clusterState, "initial-allocation");
+            assertThat(desiredBalanceAllocator.getDesiredBalance(), not(equalTo(DesiredBalance.INITIAL)));
+
+            final var shutdownType = randomFrom(Type.SIGTERM, Type.REMOVE, Type.REPLACE);
+            final var singleShutdownMetadataBuilder = SingleNodeShutdownMetadata.builder()
+                .setNodeId(node2.getId())
+                .setReason("test")
+                .setType(shutdownType)
+                .setStartedAtMillis(randomNonNegativeLong());
+            if (shutdownType.equals(Type.REPLACE)) {
+                singleShutdownMetadataBuilder.setTargetNodeName(randomIdentifier());
+            } else if (shutdownType.equals(Type.SIGTERM)) {
+                singleShutdownMetadataBuilder.setGracePeriod(TimeValue.MAX_VALUE);
+            }
+            final var nodeShutdownMetadata = new NodesShutdownMetadata(Map.of(node2.getId(), singleShutdownMetadataBuilder.build()));
+
+            clusterState = ClusterState.builder(clusterState)
+                .metadata(Metadata.builder(clusterState.metadata()).putCustom(NodesShutdownMetadata.TYPE, nodeShutdownMetadata))
+                .build();
+            ClusterServiceUtils.setState(clusterService, clusterState);
+
+            assertTrue(desiredBalanceAllocator.getProcessedNodeShutdowns().isEmpty());
+            rerouteAndWait(service, clusterState, "reroute-after-shutdown");
+            assertTrue("desired balance reset should be called on node shutdown", resetCalled.get());
+            assertThat(desiredBalanceAllocator.getProcessedNodeShutdowns(), equalTo(Set.of(node2.getId())));
+
+            resetCalled.set(false);
+            rerouteAndWait(service, clusterState, "random-reroute");
+            assertFalse("desired balance reset should not be called again for processed shutdowns", resetCalled.get());
+            assertThat(desiredBalanceAllocator.getProcessedNodeShutdowns(), equalTo(Set.of(node2.getId())));
+
+            final var clusterStateBuilder = ClusterState.builder(clusterState)
+                .metadata(Metadata.builder(clusterState.metadata()).putCustom(NodesShutdownMetadata.TYPE, NodesShutdownMetadata.EMPTY));
+            final var nodeRemovedFromCluster = randomBoolean();
```

Comment:
I would like to suggest to unconditionally do an empty reroute before the node is removed.
This way the test is more linear and simpler to read.

Reply:
Why is that important? We have a reroute at L720, does that count?

Reply:
It seems that the test is structured as following now:

```
final var nodeRemovedFromCluster = randomBoolean();
if (nodeRemovedFromCluster) {
  //remove node
}
conditionally assert reset and reroute

if (nodeRemovedFromCluster==false) {
  //remove node
  //assert reset and reroute
}
```

I think the test would be a little simpler if instead we
```
if (sometimes) {
  //do empty reroute
  //assert no reset
}
//remove node
//assert reset and reroute
```
or even do an empty reroute on every test execution

Reply:
I've simplified the last part of the test, and also clarified some names. Hope it helps! See https://github.com/elastic/elasticsearch/pull/106998/commits/e83bef4b2083918af0f278ede2d94f15f8ba75ec.